File: analyzers/health.py
"""
X-Ray LLM — Project health, remediation time, release readiness.
"""

import logging
import os
import re

from analyzers._shared import _walk_py
from xray.types import HealthResult, RemediationEstimate


def check_project_health(directory: str) -> HealthResult:
    """Check for essential project files and configuration."""
    checks = []

    def _check(name, patterns, description, severity="MEDIUM"):
        for pat in patterns:
            target = os.path.join(directory, pat)
            if os.path.exists(target):
                checks.append(
                    {"name": name, "status": "pass", "file": pat, "description": description, "severity": severity}
                )
                return
        checks.append(
            {"name": name, "status": "fail", "file": patterns[0], "description": description, "severity": severity}
        )

    _check("README", ["README.md", "README.rst", "README.txt", "README"], "Project documentation", "HIGH")
    _check("LICENSE", ["LICENSE", "LICENSE.md", "LICENSE.txt", "LICENCE", "COPYING"], "License file", "MEDIUM")
    _check(".gitignore", [".gitignore"], "Git ignore rules", "MEDIUM")
    _check(
        "Requirements",
        ["requirements.txt", "pyproject.toml", "setup.py", "setup.cfg", "Pipfile", "poetry.lock", "uv.lock"],
        "Dependency specification",
        "HIGH",
    )
    _check(
        "CI Config",
        [".github/workflows", ".gitlab-ci.yml", "Jenkinsfile", ".circleci", ".travis.yml", "azure-pipelines.yml"],
        "CI/CD configuration",
        "LOW",
    )
    _check("Tests", ["tests", "test", "tests.py", "test.py"], "Test directory or file", "HIGH")
    _check(
        "Type Hints",
        ["py.typed", "pyproject.toml", "mypy.ini", ".mypy.ini", "pyrightconfig.json"],
        "Type checking configuration",
        "LOW",
    )
    _check(
        "Linter Config",
        [".ruff.toml", "ruff.toml", "pyproject.toml", ".flake8", ".pylintrc", "tox.ini"],
        "Linter configuration",
        "LOW",
    )
    _check("Changelog", ["CHANGELOG.md", "CHANGELOG.rst", "CHANGES.md", "HISTORY.md"], "Change log", "LOW")
    _check("Editor Config", [".editorconfig"], "Editor configuration", "LOW")

    passed = sum(1 for c in checks if c["status"] == "pass")
    total = len(checks)
    score = round(passed / total * 100) if total else 0

    return {
        "score": score,
        "passed": passed,
        "total": total,
        "checks": checks,
    }


# Time estimates per rule pattern
_TIME_ESTIMATES = {
    "SEC-": {"label": "~15 min", "minutes": 15},
    "QUAL-": {"label": "~5 min", "minutes": 5},
    "PY-": {"label": "~10 min", "minutes": 10},
}


def estimate_remediation_time(findings: list) -> RemediationEstimate:
    """Estimate remediation time per finding based on rule category."""
    total_min = 0
    estimates = []
    for f in findings:
        rid = f.get("rule_id", "")
        est = {"label": "~10 min", "minutes": 10}  # default
        for prefix, val in _TIME_ESTIMATES.items():
            if rid.startswith(prefix):
                est = val
                break
        total_min += est["minutes"]
        estimates.append(est["label"])

    return {
        "total_minutes": total_min,
        "total_hours": round(total_min / 60, 1),
        "per_finding": estimates,
    }


def check_release_readiness(directory: str) -> dict:
    """Assess release readiness based on multiple criteria."""
    checks = []

    # Version in pyproject.toml
    pyproject = os.path.join(directory, "pyproject.toml")
    has_version = False
    if os.path.exists(pyproject):
        try:
            with open(pyproject, encoding="utf-8") as f:
                content = f.read()
            has_version = "version" in content.lower()
        except OSError as e:
            logging.debug("Skipped pyproject.toml version check: %s", e)
    checks.append({"name": "Version defined", "pass": has_version, "severity": "HIGH"})

    # CHANGELOG exists and is recent
    changelog_exists = any(
        os.path.exists(os.path.join(directory, f))
        for f in ["CHANGELOG.md", "CHANGELOG.rst", "CHANGES.md", "HISTORY.md"]
    )
    checks.append({"name": "Changelog exists", "pass": changelog_exists, "severity": "MEDIUM"})

    # No TODO/FIXME in critical paths
    critical_tods = 0
    for fpath, _ in _walk_py(directory):
        try:
            with open(fpath, encoding="utf-8", errors="ignore") as f:
                for line in f:
                    if re.search(r"\b(FIXME|XXX|BUG)\b", line, re.IGNORECASE):
                        critical_tods += 1
        except OSError as e:
            logging.debug("Skipped TODO scan for %s: %s", fpath, e)
    checks.append(
        {"name": f"No critical TODOs ({critical_tods} found)", "pass": critical_tods == 0, "severity": "HIGH"}
    )

    # Tests exist
    test_dir = any(os.path.isdir(os.path.join(directory, d)) for d in ["tests", "test"])
    checks.append({"name": "Tests exist", "pass": test_dir, "severity": "HIGH"})

    # README exists
    readme = any(os.path.exists(os.path.join(directory, f)) for f in ["README.md", "README.rst", "README"])
    checks.append({"name": "README exists", "pass": readme, "severity": "MEDIUM"})

    # No print() debug statements (rough check)
    debug_prints = 0
    for fpath, rel in _walk_py(directory):
        if "test" in rel.lower():
            continue
        try:
            with open(fpath, encoding="utf-8", errors="ignore") as f:
                for line in f:
                    stripped = line.strip()
                    if stripped.startswith("print(") and "debug" not in stripped.lower():
                        debug_prints += 1
        except OSError as e:
            logging.debug("Skipped debug-print scan for %s: %s", fpath, e)
    checks.append({"name": f"No debug prints ({debug_prints} found)", "pass": debug_prints < 5, "severity": "LOW"})

    # .env not committed
    env_file = os.path.join(directory, ".env")
    gitignore = os.path.join(directory, ".gitignore")
    env_safe = True
    if os.path.exists(env_file) and os.path.exists(gitignore):
        try:
            with open(gitignore, encoding="utf-8") as f:
                env_safe = ".env" in f.read()
        except OSError as e:
            logging.debug("Skipped .gitignore check: %s", e)
    checks.append({"name": ".env in .gitignore", "pass": env_safe, "severity": "HIGH"})

    passed = sum(1 for c in checks if c["pass"])
    total = len(checks)
    score = round(passed / total * 100) if total else 0

    return {
        "score": score,
        "passed": passed,
        "total": total,
        "checks": checks,
        "ready": score >= 80,
    }

File: scanner/src/analyzers/health.rs
//! Project health checks, remediation time estimation, release readiness.
//! Rust port of analyzers/health.py.

use std::collections::HashMap;
use std::path::Path;
use walkdir::WalkDir;

use crate::constants::SKIP_DIRS;

/// A single health check result.
#[derive(Debug, Clone, serde::Serialize)]
pub struct HealthCheck {
    pub name: String,
    pub status: String,
    pub file: String,
    pub description: String,
    pub severity: String,
}

/// Aggregated health result.
#[derive(Debug, Clone, serde::Serialize)]
pub struct HealthResult {
    pub score: usize,
    pub passed: usize,
    pub total: usize,
    pub checks: Vec<HealthCheck>,
}

/// Remediation time estimate.
#[derive(Debug, Clone, serde::Serialize)]
pub struct RemediationEstimate {
    pub total_minutes: usize,
    pub total_hours: f64,
    pub per_finding: Vec<String>,
}

/// Check for essential project files and configuration.
pub fn check_project_health(directory: &str) -> HealthResult {
    let mut checks = Vec::new();

    let check_items: Vec<(&str, Vec<&str>, &str, &str)> = vec![
        ("README", vec!["README.md", "README.rst", "README.txt", "README"], "Project documentation", "HIGH"),
        ("LICENSE", vec!["LICENSE", "LICENSE.md", "LICENSE.txt", "LICENCE", "COPYING"], "License file", "MEDIUM"),
        (".gitignore", vec![".gitignore"], "Git ignore rules", "MEDIUM"),
        ("Requirements", vec!["requirements.txt", "pyproject.toml", "setup.py", "setup.cfg", "Pipfile", "poetry.lock", "uv.lock"], "Dependency specification", "HIGH"),
        ("CI Config", vec![".github/workflows", ".gitlab-ci.yml", "Jenkinsfile", ".circleci", ".travis.yml", "azure-pipelines.yml"], "CI/CD configuration", "LOW"),
        ("Tests", vec!["tests", "test", "tests.py", "test.py"], "Test directory or file", "HIGH"),
        ("Type Hints", vec!["py.typed", "pyproject.toml", "mypy.ini", ".mypy.ini", "pyrightconfig.json"], "Type checking configuration", "LOW"),
        ("Linter Config", vec![".ruff.toml", "ruff.toml", "pyproject.toml", ".flake8", ".pylintrc", "tox.ini"], "Linter configuration", "LOW"),
        ("Changelog", vec!["CHANGELOG.md", "CHANGELOG.rst", "CHANGES.md", "HISTORY.md"], "Change log", "LOW"),
        ("Editor Config", vec![".editorconfig"], "Editor configuration", "LOW"),
    ];

    for (name, patterns, description, severity) in &check_items {
        let mut found = false;
        let mut found_file = patterns[0].to_string();
        for pat in patterns {
            let target = Path::new(directory).join(pat);
            if target.exists() {
                found = true;
                found_file = pat.to_string();
                break;
            }
        }
        checks.push(HealthCheck {
            name: name.to_string(),
            status: if found { "pass" } else { "fail" }.to_string(),
            file: found_file,
            description: description.to_string(),
            severity: severity.to_string(),
        });
    }

    let passed = checks.iter().filter(|c| c.status == "pass").count();
    let total = checks.len();
    let score = if total > 0 { (passed * 100) / total } else { 0 };

    HealthResult { score, passed, total, checks }
}

/// Estimate remediation time per finding based on rule category.
pub fn estimate_remediation_time(findings: &[serde_json::Value]) -> RemediationEstimate {
    let time_map: HashMap<&str, (&str, usize)> = [
        ("SEC-", ("~15 min", 15)),
        ("QUAL-", ("~5 min", 5)),
        ("PY-", ("~10 min", 10)),
        ("PORT-", ("~10 min", 10)),
    ].iter().cloned().collect();

    let mut total_min = 0;
    let mut estimates = Vec::new();

    for f in findings {
        let rule_id = f.get("rule_id").and_then(|v| v.as_str()).unwrap_or("");
        let (label, minutes) = time_map
            .iter()
            .find(|(prefix, _)| rule_id.starts_with(*prefix))
            .map(|(_, v)| *v)
            .unwrap_or(("~10 min", 10));
        total_min += minutes;
        estimates.push(label.to_string());
    }

    RemediationEstimate {
        total_minutes: total_min,
        total_hours: (total_min as f64 / 60.0 * 10.0).round() / 10.0,
        per_finding: estimates,
    }
}

/// Assess release readiness based on multiple criteria.
pub fn check_release_readiness(directory: &str) -> serde_json::Value {
    let mut checks = Vec::new();
    let dir = Path::new(directory);

    // Version in pyproject.toml
    let has_version = dir.join("pyproject.toml")
        .is_file()
        && std::fs::read_to_string(dir.join("pyproject.toml"))
            .unwrap_or_default()
            .to_lowercase()
            .contains("version");
    checks.push(serde_json::json!({"name": "Version defined", "pass": has_version, "severity": "HIGH"}));

    // Changelog exists
    let changelog = ["CHANGELOG.md", "CHANGELOG.rst", "CHANGES.md", "HISTORY.md"]
        .iter()
        .any(|f| dir.join(f).exists());
    checks.push(serde_json::json!({"name": "Changelog exists", "pass": changelog, "severity": "MEDIUM"}));

    // No critical TODOs
    let critical_re = regex::Regex::new(r"(?i)\b(FIXME|XXX|BUG)\b").unwrap();
    let mut critical_todos = 0usize;
    for entry in walk_py_files(directory) {
        if let Ok(content) = std::fs::read_to_string(&entry) {
            for line in content.lines() {
                if critical_re.is_match(line) {
                    critical_todos += 1;
                }
            }
        }
    }
    checks.push(serde_json::json!({
        "name": format!("No critical TODOs ({critical_todos} found)"),
        "pass": critical_todos == 0,
        "severity": "HIGH"
    }));

    // Tests exist
    let tests = ["tests", "test"].iter().any(|d| dir.join(d).is_dir());
    checks.push(serde_json::json!({"name": "Tests exist", "pass": tests, "severity": "HIGH"}));

    // README exists
    let readme = ["README.md", "README.rst", "README"].iter().any(|f| dir.join(f).exists());
    checks.push(serde_json::json!({"name": "README exists", "pass": readme, "severity": "MEDIUM"}));

    let passed = checks.iter().filter(|c| c["pass"].as_bool().unwrap_or(false)).count();
    let total = checks.len();
    let score = if total > 0 { (passed * 100) / total } else { 0 };

    serde_json::json!({
        "ready": passed == total,
        "score": score,
        "passed": passed,
        "total": total,
        "checks": checks,
    })
}

/// Walk Python files in a directory (utility).
fn walk_py_files(directory: &str) -> Vec<std::path::PathBuf> {
    let mut result = Vec::new();
    for entry in WalkDir::new(directory)
        .into_iter()
        .filter_entry(|e| {
            if e.file_type().is_dir() {
                let name = e.file_name().to_str().unwrap_or("");
                !SKIP_DIRS.contains(&name) && !name.starts_with('.')
            } else {
                true
            }
        })
        .filter_map(|e| e.ok())
    {
        if entry.file_type().is_file() {
            if let Some("py") = entry.path().extension().and_then(|e| e.to_str()) {
                result.push(entry.path().to_path_buf());
            }
        }
    }
    result
}
